net: only enforce expected services for half of outgoing connections

theuni opened the pull request:
This ensures that future flags can be turned off after being enabled without causing all outbound nodes to refuse to connect to it in the future for lack of expected services.

Should probably be backported to 0.14 as well.

laanwj commented:
utACK https://github.com/bitcoin/bitcoin/pull/10441/commits/0f1e7d3623f37d8ba12ea63543794c90682567bd

jonasschnelli reviewed (approved):
utACK 0f1e7d3623f37d8ba12ea63543794c90682567bd

theuni committed:
net: only enforce the services required to connect

also once half of all outgoing nodes have our preferred flags, require only
minimal flags from the rest.

gmaxwell commented:
NACK. This will result in a rather high probability (1/256) that a node ends up with no peers with the expected flags.

theuni commented:
Updated to fix an issue pointed out by @gmaxwell

sdaftuar commented:
utACK 7e1718d

morcos commented:
utACK 7e1718d

jonasschnelli commented:
(re)utACK 7e1718d389d6cb1223124cefa594f136f56ec4d1

gmaxwell commented:
Perhaps consider bracing the conditionals? (at least the new ones?)

theuni commented:
@gmaxwell done. diff from before is cosmetic only:
```diff
diff --git a/src/net.cpp b/src/net.cpp
index 1e50cfb..22d26a9 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -1796,11 +1796,13 @@ void CConnman::ThreadOpenConnections()
 
             // only consider nodes missing relevant services after 40 failed attempts and only if less than half the outbound are up.
             ServiceFlags nRequiredServices = nRelevantServices;
-            if (nTries >= 40 && nOutbound < (nMaxOutbound >> 1))
+            if (nTries >= 40 && nOutbound < (nMaxOutbound >> 1)) {
                 nRequiredServices = REQUIRED_SERVICES;
+            }
 
-            if ((addr.nServices & nRequiredServices) != nRequiredServices)
+            if ((addr.nServices & nRequiredServices) != nRequiredServices) {
                 continue;
+            }
 
             // do not allow non-default ports, unless after 50 invalid addresses selected already
             if (addr.GetPort() != Params().GetDefaultPort() && nTries < 50)
```

sipa commented on src/net.cpp:
```diff
@@ -1789,14 +1795,27 @@ void CConnman::ThreadOpenConnections()
+            // regardless of the services assumed to be available, only require the minimum if half or more outbound have relevant services
+            if (nOutboundRelevant >= (nMaxOutbound >> 1)) {
+                addrConnect.nServices = REQUIRED_SERVICES;
+            } else {
```
I think this else branch is unnecessary?

theuni replied:
It possibly downgrades the requirement from [line 1799](https://github.com/bitcoin/bitcoin/pull/10441/files#diff-9a82240fe7dfe86564178691cc57f2f1R1799).

sipa replied:
Indeed!

sipa commented:
```
net.cpp: In member function ‘void CConnman::ThreadOpenConnections()’:
net.cpp:1732:115: warning: self-comparison always evaluates to true [-Wtautological-compare]
                     if (pnode->fSuccessfullyConnected && !pnode->fFeeler && (pnode->nServices & nRelevantServices == nRelevantServices)) {
                                                                                                 ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
net.cpp:1732:115: warning: suggest parentheses around comparison in operand of ‘&’ [-Wparentheses]
```

theuni commented:
Yikes, thanks. Pushed the fixup.

TheBlueMatt commented:
utACK b6fbfc228236947eaea5c14dda299f5a01810e92
